src/kernel/retype_tbl.c

```diff
@@ -95,125 +227,235 @@ mod_mem_type(void *pa, const mem_type_t type)
+	for (i = 0; i < NUM_CPU; i++) {
+		retype_info = GET_RETYPE_CPU_ENTRY(i, idx, order);
+		old_temp.refcnt_atom.v = retype_info->refcnt_atom.v;
+		temp.refcnt_atom.v = retype_info->refcnt_atom.v;
+		temp.refcnt_atom.type = RETYPETBL_RETYPING;
+		/* See if we can successfully lock it */
+		if (retypetbl_cas(&(retype_info->refcnt_atom.v), old_temp.refcnt_atom.v, temp.refcnt_atom.v) != CAS_SUCCESS) cos_throw(err, -EINVAL);
+	}
 
-		ref_sum += local_u.ref_cnt;
+	/* Now all of them are frozen. add up the counts to see this still mapped somewhere */
+	last_unmap = 0;
+	sum = 0;
+	for (cpu = 0; cpu < NUM_CPU; cpu++) {
+		sum += GET_RETYPE_CPU_ENTRY(cpu, idx, order)->refcnt_atom.ref_cnt;
 		/* for tlb quiescence check */
-		if (last_unmap < retype_tbl[cpu].mem_set[idx].last_unmap)
-			last_unmap = retype_tbl[cpu].mem_set[idx].last_unmap;
+		if (last_unmap < GET_RETYPE_CPU_ENTRY(cpu, idx, order)->last_unmap)
+			last_unmap = GET_RETYPE_CPU_ENTRY(cpu, idx, order)->last_unmap;
 	}
-	/* only can retype when there's no more mapping */
-	if (ref_sum != 0) cos_throw(restore_all, -EINVAL);
 
-	/* before retype any type of memory back to untyped, we need
+	/* Only can retype when there's no more mapping */
+	if (sum != 0) cos_throw(err, -EINVAL);
+
+	/* 
+	 * Before retype any type of memory back to untyped, we need
 	 * to make sure TLB quiescence has been achieved after the
-	 * last unmapping of any pages in this memory set. */
-	if (!tlb_quiescence_check(last_unmap)) cos_throw(restore_all, -EQUIESCENCE);
+	 * last unmapping of any pages in this memory set
+	 */
+	if (!tlb_quiescence_check(last_unmap)) cos_throw(err, -EQUIESCENCE);
 	cos_mem_fence();
 
-	/**************************/
-	/**** Legit to retype! ****/
-	/**************************/
-	/* we already locked all entries. feel free to change here. */
+	/* Repetitively add the record of a page set to a type, into the level one above it */
+	if (old_type == RETYPETBL_USER) {
+		for (j = MAX_PAGE_ORDER; j > (int)order; j = page_sizes_small[j]) {
+			glb_retype_info = GET_GLB_RETYPE_ENTRY(idx, j);
+			/* The following is atomic with FAA */
+			cos_faa((int*)glb_retype_info, -(1 << RETYPE_ENT_GLB_REFCNT_SZ));
+		}
+	} else {
+		for (j = MAX_PAGE_ORDER; j > (int)order; j = page_sizes_small[j]) {
+			glb_retype_info = GET_GLB_RETYPE_ENTRY(idx, j);
+			/* The following is atomic with FAA */
+			cos_faa((int*)glb_retype_info, -1);
+		}
+	}
+
+	/* Update all the per-cpu variables one by one */
 	for (cpu = 0; cpu < NUM_CPU; cpu++) {
-		retype_tbl[cpu].mem_set[idx].refcnt_atom.ref_cnt = 0;
-		retype_tbl[cpu].mem_set[idx].refcnt_atom.type    = RETYPETBL_UNTYPED;
+		retype_info = GET_RETYPE_CPU_ENTRY(cpu, idx, order);
+		retype_info->refcnt_atom.v = 0;
 	}
-	cos_mem_fence();
 
-	/* and commit to the global entry. */
-	ret = retypetbl_cas(&(glb_retype_info->type), RETYPETBL_RETYPING, RETYPETBL_UNTYPED);
+	old_temp.refcnt_atom.v = glb_retype_info->refcnt_atom.v;
```

Comment:
It seems like you're taking these snapshots in a way that looks like it has the same structure each time.  Move into a function as this is redundant?